### PR TITLE
feat(storage): reduce size of compaction task

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

I found that there may be several large compaction task which larger than 2GB. To use parallel compact we will not use fast compact algorithm for those task larger than 2GB. When I seek the log file I find that some file in input of compaction is very large (similar to 512MB). I think that is because we do not cut file for those vnode with too much data.

A large compaction task will cost a long time to finish and it may also bring larger write amplification.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [x] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by Little-Wallace** [comment]
Main branch:
image

This branch:
image

We can see that the max size of task decrease a lot. But compaction throughput only decrease a little (Almost not different)

File: src/common/src/config.rs [hunk 1]
```diff
@@ -1376,7 +1376,7 @@ pub mod default {
 
         // decrease this configure when the generation of checkpoint barrier is not frequent.
         const DEFAULT_TIER_COMPACT_TRIGGER_NUMBER: u64 = 6;
-        const DEFAULT_TARGET_FILE_SIZE_BASE: u64 = 32 * 1024 * 1024; // 32MB
+        const DEFAULT_TARGET_FILE_SIZE_BASE: u64 = 16 * 1024 * 1024; // 32MB
```
Comment (Li0k) [inline src/common/src/config.rs:1379]: fix doc

**Comment by Li0k** [comment]
The changes in this PR are on the critical path:
1. reduce target_file_size (both base l0 and high level).
2. reduce the max select size of MinOverlappingPicker.
3. remove the `vnode_changed` switch condiction of SstableBuilder.

maybe we need more detailed testing and conclusions
1. Nexmark Benchmark like Q20 (I found that the performance of the branch test has slightly regressed, and the number of pending bytes has increased.)
2. Backfill test (Since we are also doing SST alignment during the backfill stage, I hope to test this path.)

**Comment by Little-Wallace** [comment]
Here is backfill benchmark result:
main branch:
image
current branch:
image


Current branch could reduce the size of compact task. But in this case, the compactor read throughput of current branch seems to be a little higher.

**Review by Li0k — APPROVED** [review]

File: src/storage/src/hummock/sstable/multi_builder.rs [hunk 2]
```diff
@@ -195,15 +188,7 @@ where
...
-                if switch_builder {
-                    need_seal_current = true;
-                } else if builder.reach_capacity() {
-                    if !self.is_target_level_l0_or_lbase || builder.reach_max_sst_size() {
-                        need_seal_current = true;
-                    } else {
-                        need_seal_current = self.is_target_level_l0_or_lbase && vnode_changed;
-                    }
-                }
+                need_seal_current = switch_builder || builder.reach_capacity();
```
Comment (Li0k) [inline src/storage/src/hummock/sstable/multi_builder.rs:191]: we can also remove `reach_max_sst_size` and `is_target_level_l0_or_lbase` ?

Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/multi_builder.rs:191]: fixed

**Pull request merged**